### ISSUE TITLE
[BUG] Long wildcard query on wildcard field causes `too_many_clauses` exception

### ISSUE THREAD

**Issue by HUSTERGS** [opensearch-project/OpenSearch#15851]
### Describe the bug

long wildcard query on wildcard field causes `too_many_clauses` exception, the `matchAllTermsQuery` function in `WildcardFieldMapper.java` directly concat all the terms generated by NGrams. Maybe the better way is to test and restrain the total clause number to a reasonable degree, so even long search strings can proceed as expected instead of throwing an exception. The only downside is verification phase may take longer time, because the qualified document may increase.

### Related component

Search

### To Reproduce

1. Create a simple index contains `wildcard` field
```
curl -X PUT "http://localhost/my-index" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "long_text": {
        "type": "wildcard"
      }
    }
  }
}
'
```
2. Generate and verify a random long string for search
```shell
MY_LONG_STRING=`openssl rand -base64 2000 | tr -dc A-Za-z0-9`
MY_LONG_STRING=${MY_LONG_STRING:0:2000}
echo ${#MY_LONG_STRING}
```
3. Search on the `wildcard` field
```
curl -X GET "http://localhost/my-index/_search" -H 'Content-Type: application/json' -d'
{
  "query": {
      "wildcard": {
          "long_text": "'"${MY_LONG_STRING}"'"
      }
  }
}
'
```

4. Get the result
```
{"error":{"root_cause":[{"type":"query_shard_exception","reason":"failed to create query: maxClauseCount is set to 1024","index":"my-index","index_uuid":"HLjKYimwRx2cVkiqzvvj5Q"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"my-index","node":"jRDKdrICRqmOdhG5fW33Pg","reason":{"type":"query_shard_exception","reason":"failed to create query: maxClauseCount is set to 1024","index":"my-index","index_uuid":"HLjKYimwRx2cVkiqzvvj5Q","caused_by":{"type":"too_many_clauses","reason":"maxClauseCount is set to 1024"}}}]},"status":400}%
```

### Expected behavior

Return the expected search result even for long search strings

**Comment by msfroh** [comment]
[Search Triage] This is a good catch! Thanks @HUSTERGS. Limiting the clause count is a good idea. We may want to keep clauses in a priority queue of limited size, where the comparator keeps the lowest-frequency terms.

@HUSTERGS -- are you able to take that on? Need any help with it?

**Comment by HUSTERGS** [comment]
Yes, I can try to fix that. When bugfix ready I'll comment under this issue @msfroh

**Comment by chamchi00800** [comment]
Hi @HUSTERGS,
 I saw your earlier comment about working on this issue. 
Are you still working on it? If not, I’d be happy to pick it up and build on any progress you’ve made.


@msfroh, 
would that be alright with you as well? 
I’m interested in exploring the approach you suggested: limiting the clause count and prioritizing lower-frequency terms.

Thank you both!

**Comment by HUSTERGS** [comment]
Hi @chamchi00800,
Thanks for reaching out, and sorry for the delayed follow-up. I'm no longer actively working on this issue, so please feel free to pick it up — go right ahead!

**Comment by msfroh** [comment]
@chamchi00800  -- yes, please!

Keep in mind, there's a mild challenge: we don't know the term frequencies at the time that we're creating the `firstPhaseQuery` (since that logic is independent of the search index). We *can* figure out term frequencies once we have an `IndexSearcher`, though, which we get in the `rewrite` method.

My suggested plan of attack (now that I've had two years away from this to clear my head) would be something like:
1. In [this branch](https://github.com/opensearch-project/OpenSearch/blob/e074f4550dc7bdb59c1c044fbad76b9de0fa4d76/server/src/main/java/org/opensearch/index/mapper/WildcardFieldMapper.java#L680-L696), instead of returning a BooleanQuery, we'd need to return a custom query type (`RequiredTermsQuery`?) that just collects all of the "required" trigrams.
2. That custom query type would implement a `rewrite` method that turns into a BooleanQuery with a bounded number of trigrams (maybe 8 or 16?), based on their frequencies. The custom query type wouldn't need to implement `getWeight` or any of the other complicated stuff -- it just exists to be rewritten. 

Since the clause count check happens right after the rewrite (literally -- [it's right at the end of the rewrite method](https://github.com/apache/lucene/blob/038e16787158a95576262842f87302b3b81032de/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L878)), we should be good.

Let me know if you need any more help!

Edit: Another caveat -- the `RequiredTermsQuery` would potentially need more complicated logic in the case-insensitive case (since each trigram can have up to 8 possible cases, and we need to sum the frequencies across those 8 cases to prioritize them). It's not impossible, but makes it all a little bit messier.

**Comment by chamchi00800** [comment]
Hello, @HUSTERGS

Thanks for getting back to me! Really appreciate it!

**Comment by chamchi00800** [comment]
Hello, @msfroh

Thanks a lot for taking the time to revisit this and for the detailed guidance!
I`ll try implementing it based on the approach you suggested.
If possible, could you also assign this issue to me?
 
Really appreciate your help, especially after all this time!
